### ISSUE TITLE
KCL: Polygon function

### ISSUE THREAD

**Issue by adamchalmers** [KittyCAD/modeling-app#4261]
Add a `polygon` function to the stdlib. It should take these parameters:

1. An object with fields {sideLength: f64, numSides: u64, center: [f64; 2]}
2. A sketchgroup

Define this in `std/shapes.rs`

**Comment by guptaarnav** [comment]
Was working on implementing this and am getting an unexpected engine error.  I am creating the polygon by computing the vertex points and extending a path to each of the lines.  My implementation of the polygon sketch works with no shown errors:
![Image](https://github.com/user-attachments/assets/bad748f3-4ec8-4c2a-adc5-ee44baef0812)


There seems to be a BadRequest API error stating "The given edge has no opposite edge" when I try to extrude on this polygon sketch, see below.  The polygon still successfully shows the extrusion as expected, but this error comes up and this same error is also causing cargo nextest tests to fail.  I wasn't able to find any documentation or discussion in github about this error except for [#3830](https://github.com/KittyCAD/modeling-app/issues/3830), which vaguely alluded to the problem being in the engine in that instance.  
![Image](https://github.com/user-attachments/assets/41eb7237-806e-4643-9234-33d70cb3fa3a).  My first guess is that this error is saying that the sketch is not closed, but it appears that it is.

**Comment by jtran** [comment]
@guptaarnav, have you tried adding `|> close(%)` on line 7 before the extrude if it's what you said about not being closed? This just came up in #4271.

**Comment by jtran** [comment]
> 1. An object with fields {sideLength: f64, numSides: u64, center: [f64; 2]}

What's the reason for using side length instead of radius? I guess I'm imagining wanting to keep a constant radius and change the number of sides. On the other hand, a side length parameter would need to be recomputed to keep the same overall size. But maybe that's not how people would actually use it?

**Comment by guptaarnav** [comment]
> [@guptaarnav](https://github.com/guptaarnav), have you tried adding `|> close(%)` on line 7 before the extrude if it's what you said about not being closed? This just came up in [#4271](https://github.com/KittyCAD/modeling-app/issues/4271). But I'd expect the polygon function to close it for you.

@jtran I already close it in the polygon function following what _circle_ does to close the sketch.  

But I tried anyways to throw an explicit _close(%)_ before the extrude in the kcl code and _close_ complained ("No such object exists of type Scene::Model::Brep::Path").  Is that confirmation that the polygon sketch is closed already?

**Comment by guptaarnav** [comment]
> > 1. An object with fields {sideLength: f64, numSides: u64, center: [f64; 2]}
> 
> What's the reason for using side length instead of radius? I guess I'm imagining wanting to keep a constant radius and change the number of sides. On the other hand, a side length parameter would need to be recomputed to keep the same overall size. But maybe that's not how people would actually use it?

I've usually seen two polygon buttons in CAD software, circumscribed polygon and inscribed polygon.  Both use radius and numSides rather than sideLength and numSides.  I agree that radius makes more sense. (also it might be weird for a user if they're messing with polygon numSides and the shape grows in correlation with the number of sides).

I think we should go with two polygon std lib functions in that case: circumscribedPolygon and inscribedPolygon ({radius: f64, numSides: u64, center: [f64; 2]})?

@adamchalmers, would you use a sideLength parameterized variant of polygon as you noted in the original Issue description?

**Comment by jtran** [comment]
> I already close it in the polygon function following what _circle_ does to close the sketch.
> 
> But I tried anyways to throw an explicit _close(%)_ before the extrude in the kcl code and _close_ complained ("No such object exists of type Scene::Model::Brep::Path"). Is that confirmation that the polygon sketch is closed already?

I think so, but not sure.

> I think we should go with two polygon std lib functions in that case: circumscribedPolygon and inscribedPolygon ({radius: f64, numSides: u64, center: [f64; 2]})?

What about ...

```
polygon({radius: f64, numSides: u64, center: [f64; 2], inscribed: bool })
```
?

**Comment by adamchalmers** [comment]
`"No such object exists of type Scene::Model::Brep::Path"` is the unfortunate error that gets returned when the object is already closed yeah.

**Comment by guptaarnav** [comment]
figured out why I was getting the "The given edge has no opposite edge" error from the API, it was because I was not setting unique id's on the lines.  Not running into error anymore!

**Comment by lf94** [comment]
Closed by [Polygon stdlib function #4300](https://github.com/KittyCAD/modeling-app/pull/4300)
